container/constructor_info.go

```diff
@@ -1,22 +1,104 @@
 package container
 
-import "reflect"
+import (
+	"reflect"
 
-// ConstructorInfo defines a special constructor type that is defined by
+	"github.com/pkg/errors"
+)
+
+// ProviderDescriptor defines a special constructor type that is defined by
 // reflection. It should be passed as a value to the Provide function.
 // Ex:
-//   option.Provide(ConstructorInfo{ ... })
-type ConstructorInfo struct {
+//   option.Provide(ProviderDescriptor{ ... })
+type ProviderDescriptor struct {
 	// In defines the in parameter types to Fn.
-	In []reflect.Type
+	In []ProviderInput
 
 	// Out defines the out parameter types to Fn.
-	Out []reflect.Type
+	Out []ProviderOutput
 
 	// Fn defines the constructor function.
-	Fn func([]reflect.Value) []reflect.Value
+	Fn func([]reflect.Value) ([]reflect.Value, error)
 
 	// Location defines the source code location to be used for this constructor
 	// in error messages.
 	Location Location
 }
+
+type ProviderInput struct {
+	Type     reflect.Type
+	Optional bool
+}
+
+type ProviderOutput struct {
+	Type reflect.Type
+}
+
+func ExtractConstructorInfo(ctr interface{}) (ProviderDescriptor, error) {
```

Comment:
what about this to be more consistent with the new naming:
```suggestion
func ExtractProviderDescriptor(pvr interface{}) (ProviderDescriptor, error) {
```
I guess we should also update the file name